### PR TITLE
Add float and double variant of tsvd

### LINKED ISSUE #347: Double variant for tSVD
We need to also handle "double" as input type for tSVD (like in KMeans and POGS).

### PR DESCRIPTION
**Pull request by navdeep-G**
* Resolves #347 
* Kept original impl of tsvd for PCA impl. Will change that in a later PR.
* Still cleaning up but wanted to open a PR

### CODE REVIEW

File: src/gpu/data/matrix.cu [hunk 1]
```diff
@@ -43,6 +73,15 @@ namespace tsvd
 		);
 	}
 
+	void multiply(Matrix<tsvd_double>& A, const tsvd_double a, DeviceContext& context)
```
Comment (mdymczyk) [inline src/gpu/data/matrix.cu]: Could these `multiply` methods be generalized into a single templated method, something like:

```
template <typename T>
void multiply(Matrix<T>& A, const T a, DeviceContext& context)
	{
		thrust::transform(A.dptr(), A.dptr() + A.size(), A.dptr(), [=]__device__ (T val)
		                  {
			                  return val * a;
		                  }
		);
	}
```
Or will the compiler complain?

Similar with other methods using `thrust` (add, subtract etc.) - anything not using `cublas`.

**Comment by mdymczyk** [comment]
Added a small comment/question but otherwise looks good 👍

File: src/gpu/data/matrix.cu (hunk 1 above)
Comment (RAMitchell) [reply src/gpu/data/matrix.cu]: You can do this it just needs to go in the header file not the .cu file.

**Comment by RAMitchell** [comment]
Looks good at a glance. As @mdymczyk mentioned you can probably remove duplicated functions in a lot of cases. Duplicate code is the root of all evil so worth some time figuring this out :)

Does the increased precision resolve the accuracy issues you were seeing before?

Also note that when you run double precision operations on a GeForce card expect them to be up to 32x slower than single precision operations.

**Comment by navdeep-G** [comment]
Thx for the comments, guys. I will look into removing duplicated code (mainly for thrust calls)

And I have to see about precision. I hope it will help.

**Comment by navdeep-G** [comment]
@mdymczyk @RAMitchell I have added templates to functions using `thrust`. Please see commit with chagne here: https://github.com/h2oai/h2o4gpu/pull/462/commits/00b8a3480bb029f94bba45f7e9970cdd2f54ff94

**Comment by navdeep-G** [comment]
@mdymczyk Not sure, but the logs show this:
```
tests_open/svd/test_tsvd_power.py::test_tsvd_power_k3 Sending interrupt signal to process

./scripts/make-docker-runtests.sh: line 32: 17087 Terminated              $docker exec ${CONTAINER_NAME} bash -c 'export HOME=`pwd`; eval "$(/root/.pyenv/bin/pyenv init -)" ; /root/.pyenv/bin/pyenv global 3.6.1; cd repo ; pip install `find /dot/src/interface_py/'${dist}' -name "*h2o4gpu-*.whl"`; pip freeze ; make '${target}
```
Will rerun as I didn't see any failures.

**Pull request merged**